aredn: Add dnsmasq DNS hostnames override for configured NTP server.

philcrump opened the pull request:
I'm putting this up in case there's alternative ideas or strong opinions on this, but marking as a draft until I've done some more testing to make sure it performs the role I'm intending, which may be useful to others as well.

We're planning (and are currently simulating across my home) a mostly-offline AREDN network with a single internet-connected node. This internet-connected node has a Raspberry Pi server, with local services including ntpd on battery-backed RTC to provide accurate time to the mesh (synced to NTP Pool when internet is available).

Client laptops and other devices connecting to other (offline) nodes in the mesh will not be able to access their default internet time servers (time.apple.com, time.windows.com, ntp.ubuntu.com, etc.), and so the idea here is to override the DNS response in dnsmasq on the nodes to return the IP of the node-configured NTP server instead (which for us is the Raspberry Pi with the RTC), allowing these devices to achieve time sync without additional client configuration.

Screenshot of the result on an Ubuntu client laptop with default ntpd configuration, connected to an offline node in the mesh:
![dns-ntp-override](https://user-images.githubusercontent.com/873295/140627928-6f01f17e-d956-476f-afb3-18d815c48444.png)

aanon4 commented:
Is there a reason to override a name rather than hand out a NTP server as a DHCP option?

philcrump commented:
That was my first plan. Unfortunately most clients do not obey the DHCP option, I can confirm that Windows and Ubuntu on my laptops do not.

KM6IAU commented:
On my network, I achieve the same without rewriting DNS requests and without relying on DHCP to handle it.

I have an EdgeRouter, and I set up a DNAT rule

```
 rule 99 {
     description "ntp://0.0.0.0 (via eth1) to 172.16.1.1"
     destination {
         group {
             port-group ntp
         }
     }
     inbound-interface eth1
     inside-address {
         address 172.16.1.1
     }
     protocol tcp_udp
     type destination
 }

```

In this case, any traffic trying to reach NTP ports originating from eth1 get NAT'd to 172.16.1.1.

aanon4 commented:
I hadn't realize just how badly that option was supported :-(

KM6IAU commented:
Yeah, I've certainly seen devices ignore the NTP server pushed by DHCP.
Plenty of devices, in fact, ignore DNS pushed by DHCP.  (My experience is that Chromecast and Google Nest devices, in particular, try to use Google DNS over DHCP's specified DNS.)

I don't think dnsmasq is the best solution to this problem.

I use a rule similar to my previous post to provide "captive DNS" as well as "captive NTP".  If a device expects to send DNS or NTP requests to the gateway, it doesn't matter what the packet's IP destination is... it is rewritten by DNAT to the IP of my choosing.

ae6xe reviewed (changes requested):
@philcrump  /etc/config.mesh has historically been used as a template to build the run time  openwrt config files in /etc/config.   hosts_ntp would better fit elsewhere, probably in /etc next to hosts.   Olsr managed hosts file has to be in /tmp so that it is modifying RAM and not flash every few minutes, to not reduce flash lifespan. In the future config.mesh might just be a .json file of unique AREDN config settings.

philcrump commented:
> On my network, I achieve the same without rewriting DNS requests and without relying on DHCP to handle it.
> I have an EdgeRouter, and I set up a DNAT rule

The issue I see with that approach is that all default configurations (that I've seen) have a hostname for the NTP, and it needs to resolve that before it'll make the request that can then be DNATed. This fails at the hostname resolution step on an non-internet-connected mesh node.

I'm aware that this proposed 'feature' is somewhat of a bodge. It's intended to improve the experience of default-configuration laptops connected to nodes as it's likely in our case that these may be owned by 3rd-parties and we may not have the ability to modify their configuration.
If it's too much of a bodge for upstream aredn then I'm fine with that, we'll keep it as a patch on top of our builds until we come up with a better solution.

> /etc/config.mesh has historically been used as a template to build the run time openwrt config files in /etc/config. hosts_ntp would better fit elsewhere, probably in /etc next to hosts

Thanks that makes sense, I'll move it into /etc/config/.

philcrump commented:
This works as intended. The text box takes a comma-delimited list, but it's also tolerant of spaces between commas and names.

If this is accepted for merge then I'm happy to add a usage explanation to the documentation repo.

ae6xe reviewed (changes requested):
/etc/mesh-release file is overwritten during the build.  The contents are from the original release this file was probably initialized.   Any contents will never make it into an image.  Let's leave this file as is please.

ae6xe reviewed (commented):
on firstboot, the /etc/hosts_override file will not yet exist, but dnsmasq.conf will have an entry looking for it.  Not sure, will this just send an error in the syslog?   Just want to make sure nothing serious occurs, e.g. like dnsmasq exits.

ae6xe commented:
could put an empty file in /etc/hosts_override on firstboot, so no errors or warnings show up.

ae6xe commented:
Suggest to add name to CONTRIBUTORS file, if not already.

philcrump commented:
> /etc/mesh-release file is overwritten during the build.
Mistakenly added, I've removed and force-pushed.

> on firstboot, the /etc/hosts_override file will not yet exist, but dnsmasq.conf will have an entry looking for it. Not sure, will this just send an error in the syslog?
I tested this and it did come up with an error, doesn't prevent dnsmasq from starting, to suppress the error I've added the empty file as you suggest.

ae6xe commented:
Thinking out loud here, and inviting comments from others before committing this change. 

Is there benefit to have a more general case where other external addresses may be desired to override with a mesh 10.x.x.x address?    Instead of a label to only override "for NTP", is it beneficial to generalize -- "override external addresses with a local mesh address"?

Upon consideration if this override IP address should be propagated across the mesh via OLSR...  while this has benefit, this seems to be too much control for one admin to change the behavior for everyone, a security concern.  common internet URLs could be redirected for all mesh users without their knowledge.  Thus, this PR's scope of impact is contained to the owner of the node (the node's LAN devices) is appropriate.

aanon4 commented:
An alternate way to do this sort of thing, especially in this off-line scenario, is to handle all this aliasing in the upstream DNS server and not have these overrides on the nodes at all. As all traffic is being sent via a single Pi here, that Pi can just become the upstream DNS and deal with the overrides there. Heck, that Pi can even intercept the 8.8.8.8 and 8.8.4.4 traffic and redirect it to its own DNS server to avoid changing node confirmations at all.

VA2XJM commented:
Not possible to test right now for me. But I must admit that NTP override
is one less thing to manage on a deployment.

Now if it can also be open on a larger scale (where we can define a list
off url and where to redirect), it might be useful for linux repository.
One can override the repos adress toward a local mirror server without
having to reconfig every machine not having access to Internet to easily
upgrade.

Le dim. 14 nov. 2021 10 h 08, Joe AE6XE ***@***.***> a écrit :

> Thinking out loud here, and inviting comments from others before
> committing this change.
>
> Is there benefit to have a more general case where other external
> addresses may be desired to override with a mesh 10.x.x.x address? Instead
> of a label to only override "for NTP", is it beneficial to generalize --
> "override external addresses with a local mesh address"?
>
> Upon consideration if this override IP address should be propagated across
> the mesh via OLSR... while this has benefit, this seems to be too much
> control for one admin to change the behavior for everyone, a security
> concern. common internet URLs could be redirected for all mesh users
> without their knowledge. Thus, this PR's scope of impact is contained to
> the owner of the node (the node's LAN devices) is appropriate.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/aredn/aredn/pull/165#issuecomment-968308255>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADGSBMICWSJEZRTHOK2OKA3UL7GG7ANCNFSM5HQFXISA>
> .
>

philcrump commented:
@aanon4 In my case there's no traffic going via the Pi, it's just running some services including an NTP server. I'd like to keep the AREDN nodes handling all the routing/DNS so that the system can operate without the Pi (just wouldn't have the NTP server which offline nodes wouldn't have anyway if not for this patch.)

I understand the controversy of this patch, I'll leave this open until the end of this weekend and then close it as rejected. I'll rebase and maintain the patch on our own builds for as long it proves to be worth the effort to us: https://github.com/swhraynet/aredn/tree/swhr
Thanks all for the comprehensive input on this and especially @ae6xe for the reviews.
